src/os/macos/lwpa/os_lock.c

```diff
@@ -0,0 +1,253 @@
+      }
+      pthread_mutex_unlock(&id->mutex);
+    }
+  }
+  return res;
+}
+
+void lwpa_signal_post(lwpa_signal_t* id)
+{
+  if (id && id->valid)
+  {
+    if (0 == pthread_mutex_lock(&id->mutex))
+    {
+      id->signaled = true;
+      pthread_cond_signal(&id->cond);
+      pthread_mutex_unlock(&id->mutex);
+    }
+  }
+}
+
+void lwpa_signal_destroy(lwpa_signal_t* id)
+{
+  if (id && id->valid)
+  {
+    pthread_cond_destroy(&id->cond);
+    pthread_mutex_destroy(&id->mutex);
+    id->valid = false;
+  }
+}
+
+bool lwpa_rwlock_create(lwpa_rwlock_t* id)
+{
+  if (id)
+  {
+    if (0 == pthread_mutex_init(&id->mutex, NULL))
+    {
+      if (0 == pthread_mutex_init(&id->readcount_mutex, NULL))
+      {
+        id->valid = true;
+        id->reader_count = 0;
+        return true;
+      }
+      else
+      {
+        pthread_mutex_destroy(&id->mutex);
+      }
+    }
+  }
+  return false;
+}
+
+bool lwpa_rwlock_readlock(lwpa_rwlock_t* id)
+{
+  bool res = false;
+  if (id && id->valid)
+  {
+    if (0 == pthread_mutex_lock(&id->mutex))
+    {
+      if (id->reader_count < MAX_READERS)
+      {
+        reader_atomic_increment(id);
+        res = true;
+      }
+      pthread_mutex_unlock(&id->mutex);
+    }
+  }
+  return res;
+}
+
+bool lwpa_rwlock_try_readlock(lwpa_rwlock_t* id)
+{
+  bool res = false;
+  if (id && id->valid)
+  {
+    if (0 == pthread_mutex_trylock(&id->mutex))
+    {
+      if (id->reader_count < MAX_READERS)
+      {
+        reader_atomic_increment(id);
+        res = true;
+      }
+      pthread_mutex_unlock(&id->mutex);
+    }
+  }
+  return res;
+}
+
+void lwpa_rwlock_readunlock(lwpa_rwlock_t* id)
+{
+  if (id && id->valid)
+    reader_atomic_decrement(id);
+}
+
+bool lwpa_rwlock_writelock(lwpa_rwlock_t* id)
+{
+  if (id && id->valid)
+  {
+    if (0 == pthread_mutex_lock(&id->mutex))
+    {
+      // Wait until there are no readers, keeping the lock so that no new readers can get in.
+      while (id->reader_count > 0)
+      {
+        usleep(1000);
+      }
+      // Hold on to the lock until writeunlock() is called
+      return true;
+    }
+  }
+  return false;
+}
+
+bool lwpa_rwlock_try_writelock(lwpa_rwlock_t* id)
+{
+  if (id && id->valid)
+  {
+    if (0 == pthread_mutex_trylock(&id->mutex))
+    {
+      // Just check once to see if there are still readers
+      if (id->reader_count > 0)
+      {
+        // Readers are present, give up the lock and return false below
+        pthread_mutex_unlock(&id->mutex);
+      }
+      else
+      {
+        // Return, holding the lock
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void lwpa_rwlock_writeunlock(lwpa_rwlock_t* id)
+{
+  if (id && id->valid)
+    pthread_mutex_unlock(&id->mutex);
+}
+
+void lwpa_rwlock_destroy(lwpa_rwlock_t* id)
+{
+  if (id && id->valid)
+  {
+    pthread_mutex_destroy(&id->mutex);
+    pthread_mutex_destroy(&id->readcount_mutex);
+    id->valid = false;
+  }
+}
+
+// TODO investigate C11 atomics for this
```

Comment:
It would be nice to use atomics rather than the mutex.  But I can see why it was a TODO

Reply:
I think creating a C11 dependency could be tricky. But I could see adding some conditional compiles to use atomics whenever available.

Reply:
In the future.  It could even be a "future" SCR.